eventlog/eventlog.py

```diff
@@ -0,0 +1,759 @@
+
+    def toJson (self) -> dict:
+        if self.blobBase is not None:
+            evt = { 'BlobBase': self.blobBase, 'BlobLength': self.blobLength }
+        else:
+            evt = self.evbuf.decode('utf-8')
+        return {
+            ** super().toJson(),
+            'Event': evt
+        }
+
+# ########################################
+# Event type: firmware blob measurement
+# ########################################
+
+class FirmwareBlobEvent (GenericEvent):
+    def __init__ (self, eventheader: Tuple, buffer: bytes, idx: int):
+        super().__init__(eventheader, buffer, idx)
+        (self.base,self.length)=struct.unpack('<QQ',buffer[idx:idx+16])
+
+    def toJson (self) -> dict:
+        return { ** super().toJson(), 'Event': {
+            'BlobBase': self.base,
+            'BlobLength': self.length
+            }}
+
+
+# ########################################
+# EFI IPL event. Used during initial program load.
+# ########################################
+# Differs from generic events in that the body of the
+# event is a zero terminated UTF-8 string describing
+# what is being loaded.
+#
+# NOTE there is a confusion about zero termination,
+# and this puts correct processing into doubt.
+# tpm2_eventlog is known to mess up and generate
+# non-UTF-8 YAML in certain cases, which yq then chokes on.
+# ########################################
+
+class EfiIPLEvent (GenericEvent):
+    def toJson (self) -> dict:
+        return {
+            ** super().toJson(),
+            'Event': { 'String': nullterm8(self.evbuf[:-1]) }
+        }
+
+# ########################################
+# Spec ID Event
+# ########################################
+# This is the first event in the log, and gets a lot of
+# special processing.
+# ########################################
+
+class SpecIdEvent (GenericEvent):
+    def __init__ (self, eventheader: Tuple, buffer: bytes, idx: int):
+        super().__init__(eventheader, buffer, idx)
+#        self.signature = uuid.UUID(bytes_le=buffer[idx:idx+16])
+        (self.signature, self.platformClass, self.specVersionMinor, self.specVersionMajor,
+         self.specErrata, self.uintnSize, self.numberOfAlgorithms) = struct.unpack('<16sIBBBBI', buffer[idx+0:idx+28])
+        idx += 28
+        self.alglist = []
+        for x in range(0, self.numberOfAlgorithms):
+            (algid, digsize) = struct.unpack('<HH', buffer[idx:idx+4])
+            idx += 4
+            self.alglist.append({
+                f'Algorithm[{x}]': None,
+                'algorithmId': Digest(algid).name,
+                'digestSize': digsize
+            })
+        (self.vendorInfoSize,)=struct.unpack('<I', buffer[idx:idx+4])
+        self.vendorInfo = buffer[idx+4:idx+4+self.vendorInfoSize]
+
+    def toJson (self):
+        j = super().toJson()
+        j.pop('DigestCount')
+        j.pop('Digests')
+        j.pop('Event')
+        j['Digest'] = self.digests[Digest.sha1].digest.hex()
+        j['SpecID'] = [{
```

Comment:
Is ok to be a list? By the name seems singular.

Reply:
I will check the documentation again to make sure. You may have a point; this is the first digest, an exception in many ways.

Reply:
I left it for now. The SPEC ID event has a single digest with a pre-determined sha1 type, but my event class has a list for all other event types ... the alternative would be to make the class object more complicated and heterogeneous.